### PR TITLE
✨ Is829/jlab login required

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

Any **template** study shared with **everyone** has a valid permalink ``/study/{project_id}``. By **default** this permalink can be used by any authenticated user (i.e. user needs to login first) but if the user wants to make it public (i.e. accessible w/o login) then it needs to request that to support. Support then will enable that by setting ``published=1`` (will change the name of the column in next PR)

NOTE: this PR needs to urgently get to staging before Wednesday

## Related issue/s

- related to ITISFoundation/osparc-issues#829

## How to test

Test with a **template** study including a jupyter and shared with **everyone**

- **Permalink** w/ login:
image
- **Permalink** w/o login: (@elisabettai please review messages in ``services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py``)
image
- **Permalink** set to **public** w/o login
image

### CODE REVIEW

**Review by sanderegg — APPROVED** [review]
I like it. But one thing I would change at some point is the error message.
I think when you click on a grafana link and you are not logged in, you get the login window automatically, then you log in and it takes you to the original link target directly.
It is the same but the feeling as a user is much nicer than seeing an error message.

**Comment by odeimaiz** [comment]
Does this mean that in order to make a permalink valid, the template needs to be shared with Everyone?

**Review by mrnicegyu11 — APPROVED** [review]
Thanks a lot for this contribution! 💪

**Review by GitHK — APPROVED** [review]
Looks ok to me. As SAN said, it this happening only for published studies, meaning only for those paths?

File: services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
-# NOTE: MSG_* strings MUST be human readable messages
+# NOTE: MSG_$(ERROR_CODE_NAME) strings MUST be human readable messages
 
-MSG_PROJECT_NOT_FOUND: Final[str] = "Cannot find any study with ID '{project_id}'."
 
+MSG_PROJECT_NOT_FOUND: Final[str] = "Cannot find any study with ID '{project_id}'"
 
-MSG_PROJECT_NOT_PUBLISHED: Final[
+
+MSG_PROJECT_NOT_PUBLISHED: Final[str] = "Cannot find any study with ID '{project_id}'"
```
Comment (elisabettai) [inline services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py:9]: I'd rephrase to: "You need to be logged in to access study with ID '{project_id}'. Please login and try again. \n If you don't have an account, write to the Support email to request one".

File: services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
+
+MSG_PUBLIC_PROJECT_NOT_PUBLISHED: Final[
     str
-] = "Cannot find any published study with ID '{project_id}'"
+] = "Cannot find any public study with ID '{project_id}'.\n Login and try again."
```
Comment (elisabettai) [inline services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py:13]: I assume this is the case where we want the study to be accessible to all users **without** login. So this error message handles the case where the box in the DB is not checked. 
If this is correct, I'd rephrase to:
"This study is not public. \n Please login and try again or contact support. \n"

**Comment by pcrespov** [comment]
> Looks ok to me. As SAN said, it this happening only for published studies, meaning only for those paths? Hopefully someone coming to `/some/other/path` will receive a different message or 404

Yes, that is what happens. ``/study`` is a special prefix path that is filtered in the reverse proxy

**Review by elisabettai — APPROVED** [review]
Hi @pcrespov, thanks a lot for this. 
I am not sure I totally understand the usage of the different types of error messages or is there only one?

For the case "**Permalink** w/o login" (where there's the panda), I'd suggest to have a message saying: 
"You need to be logged in to access study with ID '{project_id}'. \n Please login and try again. \n If you don't have an account, write to the Support email to request one."
Or we can even add the "Register" button, instead of the Support email button.

**Comment by pcrespov** [comment]
> Does this mean that in order to make a permalink valid, the template needs to be shared with Everyone?

@odeimaiz In short: Yes.
Nonetheless it is an artificial restriction that I intentionally added to simplify the logic and having it in place ASAP. This feature follows the group sharing mechanism so it can also potentially share with any other group.

File: services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py (hunk 1 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py:9]: This error happens when the linked study doe not exists or is not shared with everyone

File: services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py (hunk 2 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/studies_dispatcher/_constants.py:13]: This error happens when the linked study does not exists or is not shared with everyone or is not public ... But cannot know (due to the query) which of those conditions failed. 

NOTE: all messages include at the end ``For further information please copy this message and contact support@....``
